src/java/org/apache/cassandra/cache/InstrumentingCache.java

```diff
@@ -32,11 +32,31 @@ public class InstrumentingCache<K, V>
 
     private CacheMetrics metrics;
 
+    /**
+     * cache capacity cached for performance reasons
+     * </p>
+     * The need to cache this value arises from the fact that map.capacity() is an expensive operation for
+     * Caffeine caches, which use the cache eviction policy to determine the capacity, which, in turn,
+     * may take eviction lock and do cache maintenance before returning the actual value.
+     * </p>
+     * Cassandra frequently uses cache capacity to determine if a cache is enabled.
+     * See e.g:
+     * {@link org.apache.cassandra.db.ColumnFamilyStore#putCachedCounter},
+     * {@link org.apache.cassandra.db.ColumnFamilyStore#getCachedCounter},
+     * </p>
+     * In more stressful scenarios with many counter cache puts and gets asking the underlying cache to provide the
+     * capacity instead of using the cached value leads to a significant performance degradation.
+     * </p>
+     * No thread-safety guarantees are provided. The value may be a bit stale, and this is good enough.
```

Comment:
There are external calls that set capacity on the `ICache<K, V> map` directly (bypassing `InstrumentingCache.setCapacity`). It creates a possibility that the capacity may be invalid. Should we rather add a lightweight capacity method to ICache?